src/persist/src/s3.rs

```diff
@@ -374,6 +538,98 @@ impl Blob for S3Blob {
     }
 }
 
+#[derive(Clone, Debug)]
+struct MultipartConfig {
+    multipart_threshold: usize,
+    multipart_chunk_size: usize,
+}
+
+impl Default for MultipartConfig {
+    fn default() -> Self {
+        Self {
+            multipart_threshold: Self::DEFAULT_MULTIPART_THRESHOLD,
+            multipart_chunk_size: Self::DEFAULT_MULTIPART_CHUNK_SIZE,
+        }
+    }
+}
+
+const MB: usize = 1024 * 1024;
+const TB: usize = 1024 * 1024 * MB;
+
+impl MultipartConfig {
+    // Both of these defaults come from the official `aws cli` tool.
+    //
+    // https://github.com/aws/aws-cli/blob/2.4.14/awscli/customizations/s3/transferconfig.py#L18-L29
+    const DEFAULT_MULTIPART_THRESHOLD: usize = 8 * MB;
+    const DEFAULT_MULTIPART_CHUNK_SIZE: usize = 8 * MB;
+
+    // From https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
+    const MAX_SINGLE_UPLOAD_SIZE: usize = 5 * TB;
+    const MIN_UPLOAD_CHUNK_SIZE: usize = 5 * MB; // WIP doesn't apply to last chunk
+    const MIN_PART_NUM: u32 = 1; // WIP inclusive
+    const MAX_PART_NUM: u32 = 10_000; // WIP inclusive
+
+    fn should_multipart(&self, blob_len: usize) -> Result<bool, String> {
+        if blob_len > Self::MAX_SINGLE_UPLOAD_SIZE {
+            return Err(format!(
+                "S3 does not support blobs larger than {} bytes got: {}",
+                Self::MAX_SINGLE_UPLOAD_SIZE,
+                blob_len
+            ));
+        }
+        return Ok(blob_len > self.multipart_threshold);
+    }
+
+    fn part_iter(&self, blob_len: usize) -> MultipartChunkIter {
+        debug_assert!(self.multipart_chunk_size > MultipartConfig::MIN_UPLOAD_CHUNK_SIZE);
+        MultipartChunkIter::new(self.multipart_chunk_size, blob_len)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct MultipartChunkIter {
+    total_len: usize,
+    part_size: usize,
+    part_num: u32,
+}
+
+impl MultipartChunkIter {
+    fn new(default_part_size: usize, blob_len: usize) -> Self {
+        let max_parts: usize = usize::cast_from(MultipartConfig::MAX_PART_NUM);
+
+        // Compute the minimum part size we can use without going over the max
+        // number of parts that S3 allows. This will end up getting thrown away
+        // by the `cmp::max` for anything smaller than `max_parts *
+        // default_part_size = 80GiB`.
+        let min_part_size = (blob_len + max_parts - 1) / max_parts;
+        let part_size = cmp::max(min_part_size, default_part_size);
+        MultipartChunkIter {
+            total_len: blob_len,
+            part_size,
+            part_num: MultipartConfig::MIN_PART_NUM,
```

Comment:
I have a very very mild preference to just calling this field `part_idx` and initializing it to zero (when I read `part_num` I expect this field will be something like "the immutable number of parts this object will produce", but this field seems more like "the mutable, 1-indexed representation of the next part to be produced"

Reply:
I think `MIN_PART_NUM` encodes what we know to be the starting index for multipart uploads, and counting is 1-based here.

Reply:
I get that. I'm just saying that given those semantics it seems simpler to me to just call this field `part_idx` so we don't have to convert back to 0-indexed repr. this is a super minor nit, so feel free to tioli

Reply:
Oh yeah, sorry, I'm also for naming it `part_idx`... :sweat_smile:

Reply:
I went back and forth myself while iterating, but landed on this because part_idx wants to be a usize and part_num is a "narrower" u32. Iterating over the smaller type made it easier to be obvious about there not being overflow issues. Given this prodding, I thought about it again and realized I could keep part_idx as a u32 and convert it to a usize in just the one place, which gets us both things. It's even more obvious now that there aren't any overflow/underflow issues because this also eliminates the `-1` on an unsized type. done